REFRAME.md

```diff
@@ -0,0 +1,290 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Reframe
+
+**Author(s)**:
+- Adin Schmahmann
+- Petar Maymounkov
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+The Reframe protocol is designed for request-response messages that is sufficiently generic and extensible to evolve over time as new needs for it arise. This includes separately defining the transport and message serialization mechanisms from the actual method types and semantics.
+
+The initial use case motivating the protocol's design is to help peers discover various routing hints that enable them to find the content or service they are actually looking for. The extensibility required in this case is that various routing systems could share the protocol so that applications can work on them generically and that recursive routing systems could choose to use the protocol themselves to maximize interoperability.
+
+# Organization of this document
+
+- [Introduction](#introduction)
+- [Spec](#spec)
+- [Intended Upgrade Paths](#intended-upgrade-paths)
+- [Implementations](#implementations)
+
+# Introduction
+
+The Reframe protocol is a request-response based protocol. Upon receiving a request a Reframe server should respond to clients with information they have pertaining to the request. It's possible that a Reframe server may not have all of the information necessary for a client to actually get the data they need, but even partial resolution is acceptable. For example, if a peer is looking to download some block of data a Reframe server may only know about some peers that have the data but not their network addresses. That's ok and the client can always choose alternative mechanisms to discover the missing information. 
+
+# Spec
+
+To build an implementation of the protocol it is required to both define a transport for the messages and the method types supported by the given implementation. Over time both the method and transport types may grow and also be included in this specification.
+
+## Interaction Pattern
+
+Given that a client C wants to request information from some server S:
+
+1. C opens sends a request to S that is a single message
+2. S will respond to C with either a single message or a group of messages
+
+## Transports
+
+### HTTP + DAG-JSON
+
+All messages MUST be encoded as DAG-JSON.
+
+Requests MUST be sent as HTTP POST requests to the endpoint where the message data is sent in the body with chunked encoding.
+
+Requests and Responses MUST occur over a single HTTP call instead of the server being allowed to dial back the client with a response at a later time.
+
+If a server chooses to respond to a single request message with a group of messages in the response it should do so as a set of concatenated DAG-JSON messages (i.e. `{Response1}{Response2}...`).

```

Comment:
can these each be on a new line?

Reply:
Any reason why/why not? JSON is a self-delimiting protocol so my understanding of things is that many decoders will happily just keep on parsing once they're done with a single message.

I don't feel strongly either way, trying to understand the context.

Reply:
dag-json messages separated by `\n` sgtm.

Prior art: RPC endpoints like  /api/v0/refs/local use  http://ndjson.org (which afaik is the same/similar concept).
js-ipfs-http-client is able to [parse such response](https://github.com/ipfs/js-ipfs/blob/9a40109632e5b4837eb77a2f57dbc77fbf1fe099/packages/ipfs-http-server/test/utils/all-ndjson.js) and [yield incoming messages as soon each gets parsed](https://github.com/ipfs/js-ipfs/blob/c51b1609f9ba825b14ad1efd0b99f6af6805c4e0/packages/ipfs-http-client/src/refs/local.js#L22), which is pretty good to have when processing bigger batches.